intellij-toml/src/main/kotlin/org/toml/lang/CrateVersionReq.kt

```diff
@@ -0,0 +1,216 @@
+
+// Based on https://github.com/steveklabnik/semver/blob/master/src/version_req.rs
+class CrateVersionReq private constructor(private val predicates: List<Predicate>) {
+    companion object {
+        fun any(): CrateVersionReq =
+            CrateVersionReq(emptyList())
+
+        fun exact(version: CrateVersion) =
+            CrateVersionReq(listOf(Predicate.exact(version)))
+
+        fun parse(input: String): CrateVersionReq {
+            val requirements = input.split(",")
+            val predicates = mutableListOf<Predicate>()
+            for (requirement in requirements) {
+                val req = requirement.trim()
+                val (op, len) = when {
+                    req.startsWith(">=") -> Op.GtEq to 2
+                    req.startsWith(">") -> Op.Gt to 1
+                    req.startsWith("<=") -> Op.LtEq to 2
+                    req.startsWith("<") -> Op.Lt to 1
+                    req.startsWith("=") -> Op.Eq to 1
+                    req.startsWith("~") -> Op.Tilde to 1
+                    req.startsWith("^") -> Op.Compatible to 1
+                    req.contains('*') -> {
+                        val starPos = req.indexOf('*')
+                        val dotCount = req.dropLast(req.length - starPos).count { it == '.' }
+                        when (dotCount) {
+                            0 -> Op.WildcardMajor
+                            1 -> Op.WildcardMinor
+                            2 -> Op.WildcardPatch
+                            else -> throw IllegalArgumentException("Couldn't parse wildcard. Unrecognizable position")
+                        } to 0
+                    }
+                    else -> null to 0
+                }
+
+                val (major, minor, patch, pre, _) = IncompleteCrateVersion.parse(req.substring(len))
+                predicates += Predicate(op ?: Op.Compatible, major, minor, patch, pre)
+            }
+
+            return CrateVersionReq(predicates)
+        }
+    }
+
+    /**
+     * Is this a simple requirement that specifies a single version, such as `1.0.0` or `1.3.2`
+     */
+    val isSimple: Boolean = this.predicates.size == 1 &&
+        (this.predicates[0].op == Op.Eq ||
+            this.predicates[0].op == Op.Compatible)
+
+    fun matches(version: CrateVersion): Boolean =
+        this.predicates.isEmpty() ||
+            (this.predicates.all { it.matches(version) } &&
+                this.predicates.any { it.preTagIsCompatible(version) })
+
+    override fun toString(): String =
+        if (predicates.isEmpty()) "any"
+        else predicates.joinToString()
+}
+
+private enum class Op(val literal: String) {
+    Eq("="),
+    Gt(">"),
+    GtEq(">="),
+    Lt("<"),
+    LtEq("<="),
+    Tilde("~"),
+    Compatible("^"),
+    WildcardMajor("*"),
+    WildcardMinor(""),
+    WildcardPatch("")
+}
+
+private data class Predicate(val op: Op, private val major: Int?, private val minor: Int?, private val patch: Int?, private val pre: List<Identifier>) {
```

Comment:
Maybe use `0` instead of `null` for `major` as it does original `semver`? I think then this code will be easier to support.